### PR TITLE
[LSP] Added Symbol Renaming to the language server

### PR DESCRIPTION
**Pull request by jbylicki**
This PR implements functionality for `textDocument/prepareRename` and `textDocument/rename` capabilities.

### CODE REVIEW

File: verilog/tools/ls/symbol-table-handler.cc [hunk 1]
```diff
@@ -317,6 +342,59 @@ std::vector<verible::lsp::Location> SymbolTableHandler::FindReferencesLocations(
   return locations;
 }
 
+verible::lsp::Range SymbolTableHandler::FindRenameLocations(
```
Comment (hzeller) [inline verilog/tools/ls/symbol-table-handler.cc:345]: What comes back is not plural, but returns a single range. And it is optional.
Maybe `RenameableRangeAtCursor()` or something like that ?

File: common/lsp/lsp-protocol.yaml [hunk 2]
```diff
@@ -202,3 +202,17 @@ DocumentLinkParams:
 DocumentLink:
   range: Range
   target?: string  # DocumentUri
+
+# -- textDocument/prepareRename   (TODO(jbylicki): check that requires project + active symbol table #1189)
+PrepareRenameParams:
+  <: TextDocumentPositionParams
+
+# Response: Range[]
```
Comment (hzeller) [inline common/lsp/lsp-protocol.yaml:210]: Isn't the response just a `Range`, not `Range[]`  ( https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_prepareRename ) ?

File: verilog/tools/ls/symbol-table-handler.cc [hunk 3]
```diff
@@ -317,6 +342,59 @@ std::vector<verible::lsp::Location> SymbolTableHandler::FindReferencesLocations(
...
+    const verilog::BufferTrackerContainer &parsed_buffers) {
+  Prepare();
+  verible::LineColumnRange symbol =
+      GetTokenRangeAtTextDocumentPosition(params, parsed_buffers);
+  verible::lsp::Range range;
+  range.start.line = symbol.start.line;
+  range.start.character = symbol.start.column;
+  range.end.line = symbol.end.line;
+  range.end.character = symbol.end.column;
+  return range;
```
Comment (hzeller) [inline verilog/tools/ls/symbol-table-handler.cc:356]: What if we are over a token where a rename would not work ? The protocol definition says that in that case `null` is to be returned.
I guess that means we should return here a `std::optional<verible::lsp::Range>`  - essentially figuring out if the cursor is even hovering over something that can be rennamed (in the first simple implementation we could just look that we're not over a whitespace or keyword token, later we can look if we are actually over a symbol known to the symbol table).

And then in the `textDocument/prepareRename` request handler something that checks if the return `has_value()` and if not a nullptr.

File: verilog/tools/ls/verilog-language-server.cc [hunk 4]
```diff
@@ -138,6 +139,16 @@ void VerilogLanguageServer::SetRequestHandlers() {
         return symbol_table_handler_.FindReferencesLocations(p,
                                                              parsed_buffers_);
       });
+  dispatcher_.AddRequestHandler(
+      "textDocument/prepareRename",
+      [this](const verible::lsp::PrepareRenameParams &p) {
```
Comment (hzeller) [inline verilog/tools/ls/verilog-language-server.cc:144]: If we do the optional `lsp::Range` or `null` return here as discussed above, we might need to help out the compiler to deduce the return type by adding `-> nlohmann::json`.

File: verilog/tools/ls/verilog-language-server_test.cc [hunk 5]
```diff
@@ -1442,6 +1442,138 @@ TEST_F(VerilogLanguageServerSymbolTableTest, CheckReferenceUnknownSymbol) {
...
+               "position",
+               {{"line", params.line}, {"character", params.character}},
+           },
+           {"newName", params.newName},
+       }}};
+  return renamerq.dump();
+}
+
+// Runs tests for textDocument/rangeFormatting requests
+TEST_F(VerilogLanguageServerSymbolTableTest, PrepareRenameTest) {
```
Comment (hzeller) [inline verilog/tools/ls/verilog-language-server_test.cc:1495]: Some of this testing should probably go into `symbol-table-handler_test`

Comment (hzeller) [inline verilog/tools/ls/verilog-language-server_test.cc:1495]: The naming should reflect what we're testing, e.g. `PrepareRenameReturnsRangeOfEditableSymbol` or something along those lines. Not all tests are named like this yet, but this is helpful when seeing all the test outputs from the `bazel test //...`

File: verilog/tools/ls/verilog-language-server_test.cc [hunk 6]
```diff
@@ -1442,6 +1442,138 @@ TEST_F(VerilogLanguageServerSymbolTableTest, CheckReferenceUnknownSymbol) {
...
+  // Create sample file and make sure diagnostics do not have errors
+  RenameRequestParams params;
```
Comment (hzeller) [inline verilog/tools/ls/verilog-language-server_test.cc:1497]: Given that we're here in a test that figures out prepare rename, having the parameter called `RenameRequestParams` is of course confusing; also a good reason to switch to the auto-generated protocol structs (`verible::lsp::PrepareRenameParams`).

File: verilog/tools/ls/verilog-language-server_test.cc [hunk 7]
```diff
@@ -1442,6 +1442,138 @@ TEST_F(VerilogLanguageServerSymbolTableTest, CheckReferenceUnknownSymbol) {
...
+  EXPECT_EQ(response["result"]["start"]["line"], 2)
+      << "Invalid result for id:  ";
+  EXPECT_EQ(response["result"]["start"]["character"], 0)
+      << "Invalid result for id:  ";
+  EXPECT_EQ(response["result"]["end"]["line"], 2) << "Invalid result for id:  ";
+  EXPECT_EQ(response["result"]["end"]["character"], 3)
+      << "Invalid result for id:  ";
+}
+
+TEST_F(VerilogLanguageServerSymbolTableTest, RenameTest) {
```
Comment (hzeller) [inline verilog/tools/ls/verilog-language-server_test.cc:1530]: Multiple filenames will actually be interesting. So maybe have a test `RenameTestSymbolSingleFile` and `RenameTestSymbolMultipleFiles`, in which a symbol is defined in one file and used in another file - and the rename properly touches all files.

File: verilog/tools/ls/verilog-language-server_test.cc [hunk 8]
```diff
@@ -1442,6 +1442,138 @@ TEST_F(VerilogLanguageServerSymbolTableTest, CheckReferenceUnknownSymbol) {
...
+  ASSERT_OK(SendRequest(request));
+
+  const json response = json::parse(GetResponse());
+  std::cout << response << std::endl;
+  EXPECT_EQ(response["result"]["changes"].size(), 1)
+      << "Invalid result size for id:  ";
+  EXPECT_EQ(response["result"]["changes"][params.file].size(), 3)
+      << "Invalid result size for id:  ";
+}
+
```
Comment (hzeller) [inline verilog/tools/ls/verilog-language-server_test.cc:1576]: Additional possible tests: put the cursor over the definition of a symbol in one test and over one reference of a symbol. In both cases, the definition and all references should be updated.

Also other corner cases
  * symbol defined in a package ? Making sure that the same symbolname ('foo') defined in two different packages is not cross-renaming the other package symbol `PackageA::foo` and `PackageB::foo`

File: verilog/tools/ls/verilog-language-server_test.cc [hunk 9]
```diff
@@ -1442,6 +1442,138 @@ TEST_F(VerilogLanguageServerSymbolTableTest, CheckReferenceUnknownSymbol) {
...
+      << "textDocument/publishDiagnostics not received";
+  EXPECT_EQ(diagnostics["params"]["uri"], params.file)
+      << "Diagnostics for invalid file";
+
+  EXPECT_EQ(diagnostics["params"]["diagnostics"].size(), 0)
+      << "The test file has errors";
+  std::string request = PrepareRenameRequest(params);
+  ASSERT_OK(SendRequest(request));
+
+  const json response = json::parse(GetResponse());
```
Comment (hzeller) [inline verilog/tools/ls/verilog-language-server_test.cc:1520]: With the new potential 'null' return of this request, we should also test that.

File: verilog/tools/ls/symbol-table-handler.h [hunk 10]
```diff
@@ -65,11 +65,18 @@ class SymbolTableHandler {
       const verible::lsp::ReferenceParams &params,
       const verilog::BufferTrackerContainer &parsed_buffers);
 
+  verible::lsp::Range FindRenameLocations(
```
Comment (hzeller) [inline verilog/tools/ls/symbol-table-handler.h:68]: These new fucntions need a test in symbol-table-handler_test.cc. These functions are implicitly tested later in the verilog-language-server_test.cc, but in the symbol-table-handler_test.cc, we should check all the potential corner cases while the verilog-language-sever_test.cc is more for the final end2end testing.

An indication about the tests you can get by running the coverage by just running the specific `symbol-table-handler_test.cc` to see how much it covers from `symbol-table.cc`:

```
MODE=coverage .github/bin/build-and-test.sh verilog/tools/ls:symbol-table-handler_test
.github/bin/generate-coverage-html.sh
# Then open browser and check out the coverage of symbol-table-handler.cc
```

... Our goal is typically to be > 90% for these library specific tests.

File: common/lsp/lsp-protocol.yaml (hunk 2 above)
Comment (jbylicki) [reply common/lsp/lsp-protocol.yaml]: Yes it is! I have at one point misread the spec and implemented it for `Range[]`. It was fixed within the code itself, but I forgot it here. Changed it now

File: verilog/tools/ls/symbol-table-handler.cc (hunk 1 above)
Comment (jbylicki) [reply verilog/tools/ls/symbol-table-handler.cc]: As with the YAML spec, it is a leftover from when it did return a `Range[]`. Changed it

File: verilog/tools/ls/symbol-table-handler.cc [hunk 11]
```diff
@@ -317,6 +343,71 @@ std::vector<verible::lsp::Location> SymbolTableHandler::FindReferencesLocations(
...
+    const verilog::BufferTrackerContainer &parsed_buffers) {
+  Prepare();
+  if (files_dirty_) {
+    BuildProjectSymbolTable();
+  }
+  verible::LineColumnRange symbol =
+      GetTokenRangeAtTextDocumentPosition(params, parsed_buffers);
+  verible::lsp::Range range;
+  range = RangeFromLineColumn(symbol);
+  return range;
```
Comment (hzeller) [inline verilog/tools/ls/symbol-table-handler.cc:357]: And now, these three lines can be reduced to simply
```c++
return RangeFromLineColumn(symbol);
```

File: verilog/tools/ls/verilog-language-server_test.cc [hunk 12]
```diff
@@ -1442,6 +1442,138 @@ TEST_F(VerilogLanguageServerSymbolTableTest, CheckReferenceUnknownSymbol) {
...
 
+struct RenameRequestParams {
+  int id;
+  std::string file;
+  std::string newName;
+  int line;
+  int character;
+};
+
+std::string RenameRequest(RenameRequestParams params) {
```
[the remark this answers is not in the document]
Comment (jbylicki) [reply verilog/tools/ls/verilog-language-server_test.cc]: I have changed the methods such that they place the `(Prepare)RenameParams` into a prepared snippet of JSON. Using `to_json()` didn't make sense here, as it dumps the contents of the struct only, however the request itself also requres an ID, method and other boilerplate. Methods built into the tests already use this approach (such as `ReferencesRequest()` and I have found no logical way to construct a full request in any other manner.

File: verilog/tools/ls/verilog-language-server_test.cc (hunk 8 above)
Comment (jbylicki) [reply verilog/tools/ls/verilog-language-server_test.cc:1555]: The first test: this is known to be broken and expected to be fixed by #1678 (as well as the duplication of positions with module names); I will add the test and remove duplicate finding once that PR gets merged, if that is okay.

As for the second one: added it and it works as expected.

File: verilog/tools/ls/verilog-language-server_test.cc (hunk 7 above)
Comment (jbylicki) [reply verilog/tools/ls/verilog-language-server_test.cc]: Both are now renamed/added

File: verilog/tools/ls/symbol-table-handler.h (hunk 10 above)
Comment (jbylicki) [reply verilog/tools/ls/symbol-table-handler.h]: The new methods in this PR are now covered in `symbol-table-handler_test`

File: verilog/tools/ls/symbol-table-handler.cc [hunk 13]
```diff
@@ -317,6 +366,79 @@ std::vector<verible::lsp::Location> SymbolTableHandler::FindReferencesLocations(
...
+SymbolTableHandler::FindRenameableRangeAtCursor(
+    const verible::lsp::PrepareRenameParams &params,
+    const verilog::BufferTrackerContainer &parsed_buffers) {
+  Prepare();
+  if (files_dirty_) {
+    BuildProjectSymbolTable();
+  }
+  std::optional<verible::TokenInfo> symbol =
+      GetTokenInfoAtTextDocumentPosition(params, parsed_buffers);
+  if (symbol.has_value()) {
```
Comment (mglb) [inline verilog/tools/ls/symbol-table-handler.cc:379]:
```suggestion
  if (symbol) {
```
for consistency, as this is how you use std::optional below (`if (!location)`)

File: verilog/tools/ls/symbol-table-handler.cc [hunk 14]
```diff
@@ -317,6 +366,79 @@ std::vector<verible::lsp::Location> SymbolTableHandler::FindReferencesLocations(
...
+      GetLocationFromSymbolName(*node->Key(), node->Value().file_origin);
+  if (!location) return {};
+  std::vector<verible::lsp::Location> locations;
+  locations.push_back(location.value());
+  std::vector<verible::lsp::TextEdit> textedits;
+  CollectReferences(&root, node, &locations);
+  if (locations.empty()) return {};
+  std::map<absl::string_view, std::vector<verible::lsp::TextEdit>>
+      file_edit_pairs;
+  for (auto &loc : locations) {
```
Comment (mglb) [inline verilog/tools/ls/symbol-table-handler.cc:414]:
```suggestion
  for (const auto &loc : locations) {
```

**Review by mglb — CHANGES_REQUESTED** [review]

File: verilog/tools/ls/symbol-table-handler.h [hunk 15]
```diff
@@ -88,6 +95,15 @@ class SymbolTableHandler {
       const verible::lsp::TextDocumentPositionParams &params,
       const verilog::BufferTrackerContainer &parsed_buffers);
 
+  // TODO(jbylicki): Add docstring
```
[the remark this answers is not in the document]
Comment (jbylicki) [reply verilog/tools/ls/symbol-table-handler.h]: Done for both methods

**Review by mglb — APPROVED** [review]

**Comment by hzeller** [comment]
Looks pretty good. If you rebase, the symbol table handler has a little failure, that would probably need to be addressed.

The above coverage tool claims that only 86% of the changes have coverage, are there missed test opportunities ?

**Review by hzeller — APPROVED** [review]
Ready to submit after rebase.

**Comment by hzeller** [comment]
Rebased, then merged as #2081
